### ISSUE TITLE
[jvm-package]why the operation efficiency of xgboost4j-spark is so slow?

### ISSUE THREAD

**Issue by youngjt** [dmlc/xgboost#1176]
We have found that the operation efficiency of xgboost4j-spark is particularly slow.See the following two experiments:

The size of the data set we used is: the number of samples is about 10160000, the feature dimension of each sample is 400.

1:Stand-alone experiment:
(1)CPU:Two 8 core CPU, 32  Intel(R) Xeon(R) CPU E5-2640 v3 @ 2.60GHz,
(2)Memory:64G
**(3)training time is 41 minutes or so**

2:Cluster experiment：
(1)Specific settings for the spark-submit:：
TRAIN_PATH=/tmp/xxxx/tmp_60
TEST_PATH=/tmp/xxxxx/tmp_8
OUTPUT_PATH=/tmp/xxxxxx
numRound=100
nWorks=32
JARPATH=/home/xxxxx/xgboost/jvm-packages/xgboost4j-example/target
hadoop fs -rm -r $OUTPUT_PATH
spark-submit \ 
--class ml.dmlc.xgboost4j.scala.example.spark.DistTrainWithSpark
--master yarn-client \
--num-executors 16 \
--driver-memory 4g \
--executor-cores 2 \
--executor-memory 20g \
--queue xxxxx \
$JARPATH/xgboost4j-example-0.5-jar-with-dependencies.jar $numRound $nWorks $TRAIN_PATH $TEST_PATH $OUTPUT_PATH
**(2)training time is 92 minutes or so**

We found the "exception" information in the xgboost4j-spark distributed output log,as follows(Detailed information,see Annex1.txt)：
*_WARN Dispatcher: Message RemoteProcessDisconnected(BJHC-HADOOP-Hecate-10014.hadoop.jd.local:3493) dropped.
java.lang.IllegalStateException: RpcEnv already stopped.
    at org.apache.spark.rpc.netty.Dispatcher.postMessage(Dispatcher.scala:159)
    at org.apache.spark.rpc.netty.Dispatcher.postToAll(Dispatcher.scala:109)
    at org.apache.spark.rpc.netty.NettyRpcHandler.connectionTerminated(NettyRpcEnv.scala:630)
    at *_org.apache.spark.network.server.TransportRequestHandler.channelUnregistered(TransportRequestHandler.java:94)
    at org.apache.spark.network.server.TransportChannelHandler.channelUnregistered(TransportChannelHandler.java:89)

When our training data is changed to a biger set which has about 20000000 samples and feature dimension of each sample is 400, we get no results after a very long time(Detailed information,see Annex2.txt).

By monitoring the program execution,we have found the execution of the program stays in the 104th row of ml.dmlc.xgboost4j.scala.spark , and related code is as follows：
**102 val sparkJobThread = new Thread() {
103     override def run() {
104         boosters.foreachPartition(() => _)
105     }
106 }**

Finally,my question is:
**(1)Is this normal?
(2)How to make the xgboost4j-spark run faster?**

[Annex1.txt](https://github.com/dmlc/xgboost/files/250249/Annex1.txt)
[Annex2.txt](https://github.com/dmlc/xgboost/files/250248/Annex2.txt)

**Comment by youngjt** [comment]
![monitor](https://cloud.githubusercontent.com/assets/1442847/15043482/f4ff0570-12ff-11e6-8b94-015028b2af48.png)

**Comment by CodingCat** [comment]
from the log, I didn't see any error happened during execution, the exception is irrelevant 

would you show me the GC time of your task?

**Comment by youngjt** [comment]
We are looking for the reasons for this exception.
We think it's **rabit's **problem.It seems to be that the worker can not regist to dirver unsuccessfully.
The location of "exception" is in the function of **ReConnectLinks**.
(xgboost/rabit/src/allreduce_base.cc.
void AllreduceBase::ReConnectLinks(const char *cmd).
The problem is between 252th row and 269th row probably.)

**Comment by CodingCat** [comment]
rabit only takes are of the communication between workers and tracker, nothing to do with the exception in spark rpc layer 

I'm wondering if the copy data process between Native and Java heap space causes the GC

**Comment by youngjt** [comment]
This is the two most recent output log.

[important.out.txt](https://github.com/dmlc/xgboost/files/259414/important.out.txt)

the log of worder:
[import.part.log.txt](https://github.com/dmlc/xgboost/files/259447/import.part.log.txt)
(this file contains the information of GC.)

thanks.

**Comment by youngjt** [comment]
We have found the reason of the problem, code as follows
val partitionedData = {                           
    if (numWorkers > trainingData.partitions.length) {    
    	logger.info(s"repartitioning training set to $numWorkers partitions")
        trainingData.repartition(numWorkers)                 
    } else if (numWorkers < trainingData.partitions.length)                           
        logger.info(s"repartitioning training set to $numWorkers partitions")
        trainingData.coalesce(numWorkers)              
    } else {                                             
        trainingData                                         
    }                                                    
}
There is something wrong with the method of **trainingData.coalesce(numWorkers).**

The detailed information is as follows
(1)numWorkers=80
(2)trainingData.partitions.length=160
(3)partitionedData.partitions.length=78
the number of partitions of partitionedData is less than 80, and it leads to Rabit's communication problems.
When we have changed the functuon to **trainingData.coalesce(numWorkers, true),**everything is ok.
![error](https://cloud.githubusercontent.com/assets/1442847/15215394/e626212a-1884-11e6-93df-87910d710d79.png)

**Comment by CodingCat** [comment]
you might be right, but how can some of the experiments  move forward with less than expected number of workers?

**Comment by CodingCat** [comment]
or you have different parameter setup there?

**Comment by youngjt** [comment]
we have checked the parameters and the output log of track.py carefully, this problem is indeed.

**Comment by dirceusemighini** [comment]
Hi @youngjt It seems that this code has been replaced by:
 val partitionedData = {
      if (numWorkers != trainingData.partitions.length) {
        logger.info(s"repartitioning training set to $numWorkers partitions")
        trainingData.repartition(numWorkers)
      } else {
        trainingData
      }
    }
The @CodingCat commit says that it was to impose shuffling, but it doesn't relate to any issue.
I'm facing the same problem of slow training, but here it became slow if whenever the nworker is greater than 1. The problem is reported [here](https://github.com/dmlc/xgboost/issues/1278)
Have you been successful in increase the training performance?

**Comment by mostafa-time** [comment]
@dirceusemighini could you guys figure out the performance issue that xgboost4j-spark has? I have skimmed through other PRs, but I did not find a good answer.
I run python xgboost on a single (very large) node on ~15 millions data points with 50 features that some have very high cardinality and I get good results in just 25 minutes; however it takes it a few hours in xgboost4j-spark, and results are not as good (probably because xgboost4j-spark uses "approx" tree-method instead of "exact"?
Also I need to use the model in Spark ML pipeline and hence picked xgboost4j-spark. I was wondering if I can just convert the model obtained through python xgboost to XGBoostModel; i.e. do the training in python xgboost and convert it to XGBoostModel for scoring?

Any input is appreciated.

**Comment by CodingCat** [comment]
how many workers you have in prediction?

**Comment by mostafa-time** [comment]
@CodingCat, thanks for getting back, I was on vacation myself and just got back. Before the vacation, I changed the number of workers and the training time got reasonable so I think your hunch was right. 
I say "I think" because I don't know what do you mean by number of workers in _prediction_? Do you mean it could be that when loading python xgboost model into Spark XGBoostModel, I might get different predictions because of difference in number of workers?
